### ISSUE TITLE
[BUG] Illegal memory access when running data convert for criteo example

### ISSUE THREAD

**Issue by lazykyama** [NVIDIA/NVTabular#102]
**Describe the bug**

When running `optimize_criteo.ipynb`, I've encountered the following error. In addition, the program sometimes hangs with 100% GPU usage...

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-2af48634df31> in <module>
     47         del gdf
     48 path_out = '/data/criteo/parquet/'
---> 49 file_to_pq(train_set, 'csv', output_folder=path_out, cols=cols, dtypes=dtypes)

<ipython-input-1-2af48634df31> in file_to_pq(target_files, file_type, output_folder, cols, dtypes)
     43         if file_path != old_file_path:
     44             writer = ParquetWriter(path)
---> 45         writer.write_table(gdf)
     46         old_file_path = file_path
     47         del gdf

cudf/_lib/parquet.pyx in cudf._lib.parquet.ParquetWriter.write_table()

RuntimeError: CUDA error encountered at: /cudf/cpp/src/io/parquet/writer_impl.cu:341: 700 cudaErrorIllegalAddress an illegal memory access was encountered
```

**Steps/Code to reproduce bug**

* Download and decompress Criteo dataset (e.g., `day_0.gz` -> `day_0`)
* Launch jupyter
* Run `optimize_criteo.ipynb`

**Expected behavior**

No error and hanging-up.

**Environment details (please complete the following information):**

 - Environment location: Docker
 - Method of NVTabular install: Docker
   - NGC's container image: hash is `1567a4251e7f`.
   - Launch command: `sudo docker run --gpus=all --rm -it -v $(pwd):/ws -v /path/to/data/:/data -p 8888:8888 -p 8797:8787 -p 8796:8786 --ipc=host --cap-add SYS_PTRACE nvcr.io/nvidia/nvtabular:0.1 /bin/bash`
 - Other envs
   - Host OS: Ubuntu 18.04.4
   - GPU: TITAN X (Pascal)
   - Driver version: 440.64.00
   - Docker version: `Docker version 19.03.8, build afacb8b7f0`

**Additional context**

If needed, I can get more information by using debugging tools. Please let me know how I should do. Or, should I file this issue on `cudf` repo?

**Comment by jperez999** [comment]
@lazykyama  Sorry about hitting this issue. I would like a bit more information about your environment. Was this the first time you ran the optimize notebook. Is the output directory empty before you start? If you changed any arguments in the notebook please detail those changes as well. Does this happen as soon as you start the notebook or later in the run?

**Comment by lazykyama** [comment]
@jperez999 Yes, this error happened at the first time I ran this notebook. Precisely speaking, the first time was hang-up. So, I interrupted the notebook and tried to restart it. Then, memory error happened. In addition, I always made the output directory empty before running the notebook. I show my changes of the code at the bottom of this comment.

The timing is case-by-case. In one case, it's the first iteration of the notebook. But, in another case, the error happened after a few iterations so ~1GB file was created.

----

```diff
--- test.py.org 2020-06-23 02:02:31.226193526 +0000
+++ test.py     2020-06-23 02:12:30.197114216 +0000
@@ -1,17 +1,18 @@
 import os
-GPU_id = 4
+GPU_id = 1
 os.environ['CUDA_VISIBLE_DEVICES'] = str(GPU_id)
 import cudf
 import numpy as np
 import nvtabular as nvt
 from cudf.io.parquet import ParquetWriter

-data_path = '/datasets/criteo/crit_orig/'
+data_path = '/data/criteo/org/'
 df_train = ''
-df_valid = 'valid/'
-df_test = 'test/'
+# df_valid = 'valid/'
+# df_test = 'test/'

-train_set = [data_path + df_train + x for x in os.listdir(data_path + df_train) if x.startswith("day_") and not x.endswith("parquet")]
+train_set = [data_path + df_train + x for x in os.listdir(data_path + df_train) \
+             if x.startswith("day_") and not x.endswith("parquet")]
train_set.sort()
cont_names = ["I" + str(x) for x in range(1, 14)]
@@ -29,18 +30,20 @@

 def file_to_pq(target_files, file_type, output_folder='par_out', cols=[], dtypes=None):
     target_files.sort()
+    print(f'targets: {target_files}')
     tar = nvt.dataset(target_files, engine=file_type, names=cols, gpu_memory_frac=0.05, sep='\t', dtypes=dtypes)
     count = 0
     old_file_path = None
     writer = None
-    for gdf in tar:
+    for idx, gdf in enumerate(tar):
         #gdf.to_parquet(output_folder)
         file_path = os.path.join(output_folder, os.path.split(tar.cur_path)[1].split('.')[0])
+        print(f'[{idx}]: {file_path}')
         path = f'{file_path}.parquet'
         if file_path != old_file_path:
             writer = ParquetWriter(path)
         writer.write_table(gdf)
         old_file_path = file_path
         del gdf
-path_out = '/raid/criteo/tests/demo_out/'
+path_out = '/data/criteo/parquet/'
 file_to_pq(train_set, 'csv', output_folder=path_out, cols=cols, dtypes=dtypes)
```

```diff
--- io.py.org   2020-06-19 03:33:12.963473719 +0000
+++ /nvtabular/nvtabular/io.py  2020-06-23 01:54:45.420223091 +0000
@@ -277,7 +277,7 @@
     def __len__(self):
         return len(self.engine)

-    def set_dtypes(self, chunk):
+    def _set_dtypes(self, chunk):
         for col, dtype in self.dtypes.items():
             if type(dtype) is str:
                 if "hex" in dtype:
@@ -339,9 +339,11 @@
             raise ValueError("len(paths) must be > 0.")
         self.paths = paths
         self.kwargs = kwargs
+        self.cur_path = None

     def __iter__(self):
         for path in self.paths:
+            self.cur_path = path
             yield from GPUFileIterator(path, **self.kwargs)
```

**Comment by lazykyama** [comment]
I've applied `cuda-memcheck` to this example code. In the case of `initcheck` and `racecheck`, too many error and warning messages are displayed. I put a few messages here. If you need whole messages, please let me know. (But, especially, `initcheck` output file size is 42MB...)

### `racecheck`

Command: `cuda-memcheck --tool racecheck python test.py`
(Note that the code ran for too long time so I interrupted it...)

```
... (skip) ...
========= WARN: Race reported between Read access at 0x00003328 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int)
=========     and Write access at 0x00003110 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int) [253913 hazards]
=========
========= ERROR: Race reported between Read access at 0x00000bb0 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int)
=========     and Write access at 0x00000918 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int) [559712 hazards]
=========
========= WARN: Race reported between Read access at 0x00003328 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int)
=========     and Write access at 0x00003110 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int) [254450 hazards]
=========
========= ERROR: Race reported between Read access at 0x00000bb0 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int)
=========     and Write access at 0x00000918 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int) [561348 hazards]
=========
========= WARN: Race reported between Read access at 0x00003328 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int)
=========     and Write access at 0x00003110 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int) [256688 hazards]
=========
========= ERROR: Race reported between Read access at 0x00000bb0 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int)
=========     and Write access at 0x00000918 in cudf::io::parquet::gpu::gpuInitPageFragments(cudf::io::parquet::gpu::PageFragment*, cudf::io::parquet::gpu::EncColumnDesc const *, int, int, unsigned int, unsigned int) [567555 hazards]
=========
... (skip) ...
```

### `initcheck`

Command: `cuda-memcheck --tool initcheck --save ./initcheck.log python test.py > stdout_initcheck.log`
(Note that the following message is just one example...)

```
========= Uninitialized __global__ memory read of size 4
=========     at 0x000001b0 in void thrust::cuda_cub::cub::RadixSortScanBinsKernel<thrust::cuda_cub::cub::DeviceRadixSortPolicy<unsigned long, thrust::cuda_cub::cub::NullType, int>::Policy700, int>(thrust::cuda_cub::cub::NullType*, int)
=========     by thread (137,0,0) in block (0,0,0)
=========     Address 0x7f35b7261724
=========     Saved host backtrace up to driver entry point
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so.1 (cuLaunchKernel + 0x346) [0x2af0b6]
=========     Host Frame:/conda/envs/rapids/lib/libcudart.so.10.2 [0x1a547]
=========     Host Frame:/conda/envs/rapids/lib/libcudart.so.10.2 [0x1a5d7]
=========     Host Frame:/conda/envs/rapids/lib/libcudart.so.10.2 (cudaLaunchKernel + 0x225) [0x51885]
=========     Host Frame:/conda/envs/rapids/lib/libcudf.so (_ZN6thrust8cuda_cub3cub23RadixSortScanBinsKernelINS1_21DeviceRadixSortPolicyImNS1_8NullTypeEiE9Policy700EiEEvPT0_i + 0xdf) [0x1a1317f]
=========     Host Frame:/conda/envs/rapids/lib/libcudf.so (_ZN6thrust8cuda_cub3cub15DeviceRadixSort8SortKeysImEE9cudaErrorPvRmRNS1_12DoubleBufferIT_EEiiiP11CUstream_stb + 0x2a98) [0x1a1ac88]
=========     Host Frame:/conda/envs/rapids/lib/libcudf.so (_ZN6thrust8cuda_cub4sortINS_6detail22execute_with_allocatorIRN3rmm2mr16thrust_allocatorIcEENS0_22execute_on_stream_baseEEEPmNS_4lessImEEEEvRNS0_16execution_policyIT_EET0_SI_T1_ + 0x100) [0x1a1ea70]
=========     Host Frame:/conda/envs/rapids/lib/libcudf.so (_ZN4cudf12experimental2io6detail3csv6reader4impl18gather_row_offsetsEPKcmmP11CUstream_stPKN3rmm13device_bufferE + 0x1ec) [0x1a2d42c]
=========     Host Frame:/conda/envs/rapids/lib/libcudf.so (_ZN4cudf12experimental2io6detail3csv6reader4impl4readEmmiiiP11CUstream_st + 0x707) [0x1a2ddd7]
=========     Host Frame:/conda/envs/rapids/lib/libcudf.so (_ZN4cudf12experimental2io6detail3csv6reader15read_byte_rangeEmmP11CUstream_st + 0x2a) [0x1a2f2ca]
=========     Host Frame:/conda/envs/rapids/lib/libcudf.so (_ZN4cudf12experimental2io8read_csvERKNS1_13read_csv_argsEPN3rmm2mr22device_memory_resourceE + 0xd3a) [0x1abaada]
=========     Host Frame:/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.14.0a0+4291.g6bae03d40-py3.7-linux-x86_64.egg/cudf/_lib/csv.cpython-37m-x86_64-linux-gnu.so [0x2ec39]
=========     Host Frame:python (_PyMethodDef_RawFastCallKeywords + 0x306) [0x165286]
=========     Host Frame:python (_PyCFunction_FastCallKeywords + 0x21) [0x165301]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x5430) [0x1d2230]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x1158f9]
=========     Host Frame:python (_PyFunction_FastCallDict + 0x400) [0x116c60]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1f6a) [0x1ced6a]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0xab8) [0x1160b8]
=========     Host Frame:python (_PyFunction_FastCallKeywords + 0x387) [0x1649e7]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x14e7) [0x1ce2e7]
=========     Host Frame:python [0x16c854]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0xaa6) [0x1cd8a6]
=========     Host Frame:python (_PyGen_Send + 0x2a2) [0x16d212]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x1add) [0x1ce8dd]
=========     Host Frame:python [0x16c854]
=========     Host Frame:python [0x17744f]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0xaa6) [0x1cd8a6]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x1158f9]
=========     Host Frame:python (_PyFunction_FastCallKeywords + 0x387) [0x1649e7]
=========     Host Frame:python (_PyEval_EvalFrameDefault + 0x14e7) [0x1ce2e7]
=========     Host Frame:python (_PyEval_EvalCodeWithName + 0x2f9) [0x1158f9]
=========     Host Frame:python (PyEval_EvalCodeEx + 0x44) [0x116824]
=========     Host Frame:python (PyEval_EvalCode + 0x1c) [0x11684c]
=========     Host Frame:python [0x22c684]
=========     Host Frame:python (PyRun_FileExFlags + 0xa1) [0x236ab1]
=========     Host Frame:python (PyRun_SimpleFileExFlags + 0x1c3) [0x236ca3]
=========     Host Frame:python [0x237d95]
=========     Host Frame:python (_Py_UnixMain + 0x3c) [0x237ebc]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xe7) [0x21b97]
=========     Host Frame:python [0x1dcbd0]
=========
```

**Comment by jperez999** [comment]
@lazykyama if you could please run the following in your container and let us know of any failures you hit. `pytest /cudf/python/cudf/cudf/tests` You will be running the testing framework for cudf.

**Comment by lazykyama** [comment]
@jperez999 I ran `pytest /cudf/.../tests` and got no clear failure. At the bottom, `= 20932 passed, 923 skipped, 280 xfailed, 7 xpassed, 2703 warnings in 510.57s (0:08:30) =` was shown. `xpassed` cases might be problematic. Let me put the detail messages for these cases below.

```
XPASS ../cudf/python/cudf/cudf/tests/test_dataframe.py::test_isin_string[values4-data0] pandas's failure here seems like a bug given the reverse succeeds
XPASS ../cudf/python/cudf/cudf/tests/test_dataframe.py::test_isin_string[values4-data1] pandas's failure here seems like a bug given the reverse succeeds
XPASS ../cudf/python/cudf/cudf/tests/test_dataframe.py::test_isin_string[values4-data2] pandas's failure here seems like a bug given the reverse succeeds
XPASS ../cudf/python/cudf/cudf/tests/test_datetime.py::test_datetime_unique[some-data0]
XPASS ../cudf/python/cudf/cudf/tests/test_datetime.py::test_datetime_unique[some-data2]
XPASS ../cudf/python/cudf/cudf/tests/test_joining.py::test_merge_left_index_zero left_on/right_on produces undefined results with 0index and is disabled
XPASS ../cudf/python/cudf/cudf/tests/test_setitem.py::test_dataframe_setitem_scaler_keyerror
```
(I got these messages by additional runtime option like `pytest -rfExX -v /cudf/python/cudf/cudf/tests`. Also, I have the whole log on running the tests.)

**Comment by jperez999** [comment]
Let's go ahead and file an issue with cudf team. Unfortunately have not been able to repro due to resource limitations.

**Comment by OlivierNV** [comment]
@lazykyama Would it be possible to include the output of cuda-memcheck without the racecheck/initcheck option ? (just to capture the actual fault)

**Comment by lazykyama** [comment]
@OlivierNV Did you mean the command should be changed from, for example, `cuda-memcheck --tool racecheck python test.py` to `cuda-memcheck python test.py`? If so, I got no error from `cuda-memcheck` like below. Unfortunately, without `cuda-memcheck`, the error happened again...

```
========= CUDA-MEMCHECK
========= ERROR SUMMARY: 0 errors
```

Full console outputs are described below.

### stdout

Note that I added several progress logs.

```
targets: ['/data/criteo/org/day_0', '/data/criteo/org/day_1']
[0]: /data/criteo/parquet/day_0
         2020-07-01 23:14:23.470198 duration=4.284763051196933[sec.]
[1]: /data/criteo/parquet/day_0
         2020-07-01 23:14:50.606910 duration=4.158812602981925[sec.]
[2]: /data/criteo/parquet/day_0
         2020-07-01 23:15:17.779998 duration=4.287690730765462[sec.]
[3]: /data/criteo/parquet/day_0
         2020-07-01 23:15:45.152566 duration=4.260857192799449[sec.]
[4]: /data/criteo/parquet/day_0
... (similar logs repeated) ...
[164]: /data/criteo/parquet/day_1
         2020-07-02 00:35:11.787064 duration=4.206702688708901[sec.]
[165]: /data/criteo/parquet/day_1
         2020-07-02 00:35:28.438073 duration=1.9731408972293139[sec.]
========= CUDA-MEMCHECK
========= ERROR SUMMARY: 0 errors
```

### stderr

```
/conda/envs/rapids/lib/python3.7/site-packages/numba/cuda/envvars.py:17: NumbaWarning:
Environment variables with the 'NUMBAPRO' prefix are deprecated and consequently ignored, found use of NUMBAPRO_NVVM=/usr/local/cuda/nvvm/lib64/libnvvm.so.

For more information about alternatives visit: ('http://numba.pydata.org/numba-doc/latest/cuda/overview.html', '#cudatoolkit-lookup')
  warnings.warn(errors.NumbaWarning(msg))
/conda/envs/rapids/lib/python3.7/site-packages/numba/cuda/envvars.py:17: NumbaWarning:
Environment variables with the 'NUMBAPRO' prefix are deprecated and consequently ignored, found use of NUMBAPRO_LIBDEVICE=/usr/local/cuda/nvvm/libdevice/.

For more information about alternatives visit: ('http://numba.pydata.org/numba-doc/latest/cuda/overview.html', '#cudatoolkit-lookup')
  warnings.warn(errors.NumbaWarning(msg))
```

**Comment by kkraus14** [comment]
This should have been fixed by https://github.com/rapidsai/cudf/pull/5672. It would be great if you can confirm the fix by using the latest 0.15 nightly (once it's published from that PR).

**Comment by lazykyama** [comment]
Thanks for fixing it! I've confirmed this issue has been resolved.
(Unfortunately, the version v0.1.0 of nvtabular that this issue happened doesn't seem to support cudf v0.15. So, I confirmed the fix by nvtabular's master branch which supports cudf v0.15...)
